### PR TITLE
Rework example and e2e test script

### PR DESCRIPTION
**Pull request by TimZaman**
Submitted as WIP, everything working, just wrapping up my tests for the mpi backend. Let me know if this is OKAY. There is quite some redundancy/small bugs/outdated/unmaintained in the examples dir, so I decided to take a stab to clean it up! :-)

(Also I don't think maybe the MPI backend should be promoted, the kubeflow/MPI-operator is superior for this purpose: tailored for MPI and topology awareness).

- Unify many copies of mnist training scripts into a single one.
  - Backends: Gloo, NCCL, MPI.
  - Bases of the official pytorch/examples/mnist example.
  - Uses TensorboardX for summary writing.
  - Uses a backend argument to select backend (if applicable).
- Update all examples and dockerfiles to Pytorch 1.0.
- Remove examples using deprecated TCP backend.
- Remove old v1alpha1 examples.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pytorch-operator/126)

### CODE REVIEW

**Comment by johnugeorge** [comment]
@TimZaman Great work. Thanks for your contributions

tcp-mnist image is currently built and run during the CI process. Since the tcp example is removed, you need to point  to the new example for CI process to pass. 
   Image is built in
   https://github.com/kubeflow/pytorch-operator/blob/master/scripts/build.sh#L49
   The image which is built, is used in tests 
     https://github.com/kubeflow/pytorch-operator/tree/master/scripts/v1alpha2
     https://github.com/kubeflow/pytorch-operator/tree/master/scripts/v1beta1
      

/cc @andreyvelich 
/cc @Akado2009

File: examples/mnist/mnist.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if batch_idx % args.log_interval == 0:
+            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
```
Comment (johnugeorge) [inline examples/mnist/mnist.py]: Katib image(used in example https://github.com/kubeflow/katib/blob/master/examples/pytorchjob-example.yaml#L46 )  is currently built from https://github.com/kubeflow/pytorch-operator/tree/master/examples/katib/
     The stdout format has to follow metricname=metricvalue format.  Can you keep the same print statement as given in current code https://github.com/kubeflow/pytorch-operator/blob/master/examples/katib/dist_mnist_with_summary.py#L100

File: examples/mnist/mnist.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        for data, target in test_loader:
+            data, target = data.to(device), target.to(device)
+            output = model(data)
+            test_loss += F.nll_loss(output, target, reduction='sum').item() # sum up batch loss
+            pred = output.max(1, keepdim=True)[1] # get the index of the max log-probability
+            correct += pred.eq(target.view_as(pred)).sum().item()
+
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
```
Comment (johnugeorge) [inline examples/mnist/mnist.py]: Same comment as before

File: examples/mnist/mnist.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        datasets.MNIST('../data', train=False, transform=transforms.Compose([
+                           transforms.ToTensor(),
+                           transforms.Normalize((0.1307,), (0.3081,))
+                       ])),
+        batch_size=args.test_batch_size, shuffle=False, **kwargs)
+
+    model = Net().to(device)
+
+    if is_distributed():
+        Distributor = nn.parallel.DistributedDataParallel if args.no_cuda \
```
Comment (andreyvelich) [inline examples/mnist/mnist.py]: Why if you don't have cuda you are using `DistributedDataParallel` instead of `DistributedDataParallelCPU` ?
Is it correct that `args.no_cuda` determines that you are using GPU or not ?

File: examples/mnist/mnist.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    writer = SummaryWriter(args.dir)
+
+    torch.manual_seed(args.seed)
+
+    device = torch.device("cuda" if use_cuda else "cpu")
+
+    if should_distribute():
+        print('Using distributed PyTorch with {} backend'.format(args.backend))
+        dist.init_process_group(backend=args.backend)
```
Comment (andreyvelich) [inline examples/mnist/mnist.py]: Did you control backends for DDP? 
According to Pytorch documentation,
[DistributedDataParallelCPU](https://pytorch.org/docs/master/nn.html#torch.nn.parallel.DistributedDataParallelCPU) supports only `mpi` and `gloo` backend 
[DistributedDataParallel](https://pytorch.org/docs/master/nn.html#torch.nn.parallel.DistributedDataParallel) supports only `gloo` and `nccl`  backend. 
I might be wrong.

File: examples/mnist/mnist.py (hunk 3 above)
Comment (TimZaman) [reply examples/mnist/mnist.py]: Good catch, typo: should be `.. if use_cuda ..`

File: examples/mnist/mnist.py (hunk 4 above)
Comment (TimZaman) [reply examples/mnist/mnist.py]: If those are inconsistent/unsupported that should throw a descriptive error from the backend.

**Comment by TimZaman** [comment]
Ok thanks guys, I'll take it that this PR is appreciated (it does update a ton of stuff in examples so wasn't sure). I'll move forward to address comments and remove the WIP state.

**Comment by TimZaman** [comment]
All comments addressed. As I'm unfamiliar with this project's CI process, I'll see how things go wrt the `build.sh`.

**Comment by johnugeorge** [comment]
@TimZaman CI tests fail because mnist tests have failed. 

https://github.com/kubeflow/pytorch-operator/blob/master/scripts/v1alpha2/run-defaults.sh#L49 
https://github.com/kubeflow/pytorch-operator/blob/master/scripts/v1alpha2/run-cleanpodpolicy-all.sh#L50

Each test have a timeout of 10 minutes. I think, tests fail because they won't get completed in 10 minutes and they time out. Since we are not concerned about the accuracy of the example, we can keep the default epoch of the example to be 1(https://github.com/kubeflow/pytorch-operator/pull/126/files#diff-f5284f418f68386cd925f0081ba5dcc9R83).  WDYT?

**Comment by TimZaman** [comment]
SGTM, how do you know its timeouts?

**Comment by johnugeorge** [comment]
Timeout is a parameter in the test. 
https://github.com/kubeflow/pytorch-operator/blob/master/test/e2e/v1alpha2/defaults.go#L33

**Comment by johnugeorge** [comment]
Tests passed. Thanks @TimZaman 

/lgtm
/approve
